engines/python/setup/djl_python/scheduler/lm_block.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python
+#
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file
+# except in compliance with the License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS"
+# BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied. See the License for
+# the specific language governing permissions and limitations under the License.
+
+from abc import ABC, abstractmethod
+from transformers import GPT2LMHeadModel
+from typing import List, Dict
+
+import torch
+
+
+class GPTConfig:
+
+    def __init__(self):
+        self.numAttentionHeads = 12
+        self.numLayers = 12
+        self.hiddenStateDim = 768
+        self.logitsDim = 50257
+        self.kvDim = 64
+
+
+class LMBlock(ABC):
+
+    @abstractmethod
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def forward(self, input, past_key_values):
+        pass
+
+
+class PtLMBlock(LMBlock):
+
+    def __init__(self, model_urls: List[str], gpt_config: GPTConfig):
+        super(PtLMBlock, self).__init__()
+        self.blocks = [torch.jit.load(url) for url in model_urls]
+        self.gpt_config = gpt_config
+
+    def forward(self, input: List[torch.tensor], past_key_values):
+        return self.blocks[0](
+            *input) if past_key_values is None else self.blocks[1](
+                *input, past_key_values)
+
+
+class OrtLMBlock(LMBlock):
+
+    def __init__(self, model_urls: List[str], gpt_config: GPTConfig):
+        super(OrtLMBlock, self).__init__()
+        self.blocks = [torch.jit.load(url) for url in model_urls]
+        self.gpt_config = gpt_config
+
+    def forward(self, input: List[torch.tensor], past_key_values):
+        raise ("Not implemented yet")
+
+
+class HuggingfaceGTP2Block(LMBlock):
```

Comment:
why we want to rewrite the GPT2 Block?

Reply:
I think Qing has a point. 
Other backengines like DS, model.forward is called this way
```
model = deepspeed.init_inference(GPT2LMHeadModel.pretrained("gpt2"), config) # returns the DS engine wrapped model
model.forward(input_ids, attention_mask, ...) # this one internally calls the HF model.forward anyway
```
https://github.com/microsoft/DeepSpeed/blob/d1c3c0df5318cbcb2b221d38d4127dd566f4984d/deepspeed/inference/engine.py#L611
Not sure about FT, not sure when will their model.forward will be exposed. cc @rohithkrn  

In these cases, we might not need LMBlock itself. We could get the model and kwargs directly in SeqBatchScheduler constructor and call the model.forward directly for now.  (just thinking aloud)

And also do we need PtBlock and OrtBlock for Python engine?

Reply:
I had a discussion with @zachgk. One more thing to consider is that, this LMBlock tries to standardize the forward function but it also loads the model. In future if we introduce more models, we could use AutoModel to load, but for Peft models, the model loading by loading the base model first and then their lora updated parameters like below. 
```
peft_model_id = "mrm8488/Alpacoom"
config = PeftConfig.from_pretrained(peft_model_id)
model = AutoModelForCausalLM.from_pretrained(config.base_model_name_or_path, return_dict=True, load_in_8bit=True, device_map={"":0})
model = PeftModel.from_pretrained(model, peft_model_id)
```
Since model loading can be arbitrary as well, I am not sure whether we should have the model loading tied to the LMBlock. I don't have any hard decision comment on this, just laying out the different sides of facts.